java/src/com/google/idea/blaze/java/run/hotswap/BlazeReloadFileAction.java

```diff
@@ -0,0 +1,188 @@
+package com.google.idea.blaze.java.run.hotswap;
+
+import com.google.common.collect.ImmutableList;
+import com.google.idea.blaze.base.build.BlazeBuildService;
+import com.google.idea.blaze.base.command.buildresult.LocalFileOutputArtifact;
+import com.google.idea.blaze.base.model.primitives.Label;
+import com.google.idea.blaze.base.sync.aspects.BlazeBuildOutputs;
+import com.google.idea.blaze.base.sync.aspects.BuildResult;
+import com.google.idea.blaze.base.targetmaps.SourceToTargetMap;
+import com.intellij.debugger.DebuggerManagerEx;
+import com.intellij.debugger.impl.DebuggerSession;
+import com.intellij.debugger.impl.HotSwapFile;
+import com.intellij.debugger.impl.HotSwapManager;
+import com.intellij.debugger.settings.DebuggerSettings;
+import com.intellij.debugger.ui.HotSwapProgressImpl;
+import com.intellij.debugger.ui.HotSwapUIImpl;
+import com.intellij.debugger.ui.RunHotswapDialog;
+import com.intellij.history.core.Paths;
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.actionSystem.CommonDataKeys;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.ProjectFileIndex;
+import com.intellij.openapi.vfs.VirtualFile;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Future;
+import java.util.jar.JarInputStream;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+
+public class BlazeReloadFileAction extends AnAction {
+
+    public static final String ACTION_ID = "Debugger.ReloadFile";
```

Comment:
looks like quite generic name, will it be scoped to Bazel in some way?

Reply:
This is the JB action id that we are overriding.
https://github.com/JetBrains/intellij-community/blob/2278d05ac40658a82fd3780a4145a292898b70b2/resources/src/idea/JavaActions.xml#L514

Reply:
I think then we might need to override `CompileFile` and `CompileProject` actions as well. They doesn't work with Bazel currently.

Reply:
We do not need that as part of the PR.

Reply:
Agree, let's start with this and think about the rest of them after this one merged.